source/server-discovery-and-monitoring/server-monitoring.rst

```diff
@@ -674,13 +675,21 @@ The event API here is assumed to be like the standard `Python Event
             wait()
 
     def setUpConnection():
+        # Take the mutex to avoid a data race becauase this code writes to the connection field and a concurrent
+        # cancelCheck call could be reading from it.
+        mutex.lock()
         connection = new Connection(serverAddress)
         set connection timeout to connectTimeoutMS
-        perform connection handshake
+        mutex.unlock()
+
+        # Do any potentially blocking operations after releasing the mutex.
+        create the socket and perform connection handshake
```

Comment:
Changed from `perform connection handshake` to this because the dialing of the socket should also be done outside the lock.

Reply:
Acknowledged. In python we often say "Do not perform I/O while holding the lock."